naga/src/back/spv/writer.rs

```diff
@@ -135,6 +138,62 @@ impl Writer {
         self.capabilities_used.insert(spirv::Capability::Shader);
     }
 
+    #[allow(clippy::too_many_arguments)]
+    pub(super) fn promote_access_expression_to_variable(
+        &mut self,
+        ir_types: &UniqueArena<crate::Type>,
+        result_type_id: Word,
+        container_id: Word,
+        container_resolution: &TypeResolution,
+        index_id: Word,
+        element_ty: Handle<crate::Type>,
+        block: &mut Block,
+    ) -> Result<(Word, LocalVariable), Error> {
+        let container_type_id = self.get_expression_type_id(container_resolution);
+        let pointer_type_id = self.id_gen.next();
+        Instruction::type_pointer(
+            pointer_type_id,
+            spirv::StorageClass::Function,
+            container_type_id,
+        )
+        .to_words(&mut self.logical_layout.declarations);
```

Comment:
It turns out that, if the thing being indexed is a `TypeInner::Array`, that will never be a `TypeResolution::Local`, there will always be a `Handle`. So if we change this to expect a `Handle<Type>` instead of `container_resolution`, we can use `get_pointer_id` here as well.

I have a change that does this (and documents `TypeResolution::Local` a bit better) which I'll push to the branch tomorrow.